components/common-go/baseserver/testing.go

```diff
@@ -43,23 +43,37 @@ func MustUseRandomLocalAddress(t *testing.T) *ServerConfiguration {
 	}
 }
 
+const (
+	maxRetries = 5
+	retryDelay = 100 * time.Millisecond
+)
+
 func MustFindFreePort(t *testing.T) int {
 	t.Helper()
 
-	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
-	if err != nil {
-		t.Fatalf("cannot find free port: %v", err)
-		return 0
-	}
+	for i := 0; i < maxRetries; i++ {
+		addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+		if err != nil {
+			if i < maxRetries-1 {
+				time.Sleep(retryDelay)
+			}
+			continue
+		}
+
+		l, err := net.ListenTCP("tcp", addr)
+		if err != nil {
+			if i < maxRetries-1 {
+				time.Sleep(retryDelay)
+			}
+			continue
+		}
+		defer l.Close()
 
-	l, err := net.ListenTCP("tcp", addr)
-	if err != nil {
-		t.Fatalf("cannot find free port: %v", err)
-		return 0
+		return l.Addr().(*net.TCPAddr).Port
 	}
-	defer l.Close()
 
-	return l.Addr().(*net.TCPAddr).Port
+	t.Fatalf("cannot find free port after %d retries", maxRetries)
+	return 0
 }
```

Comment:
I suspect that we could in theory replace the whole thing with

```
 l, err := net.Listen("tcp", ":0")
    if err != nil {
        panic(err)
    }
```

I suspect the race happens for the following:
1. We try to resolve a free port,
2. We then create a listener to ensure it's open
3. We then return the port number
4. In defer, we actually close our listener
5. The parent caller then attempts to bind that port again.

But depending on how quickly it actually binds, another process could grab that port. So I think the fix here is to even avoid finding a free-port, and instead "just claim" the port directly.

In theory, it may be enough to always return 0 from this function as `0` indicates the OS should allocate the port.

Reply:
I've tried that in https://github.com/gitpod-io/gitpod/pull/17459 to see how CI reacts to it. There may be more changes needed